frontend/src/components/resourceMap/graph/graphFiltering.test.ts

```diff
@@ -0,0 +1,74 @@
+import { KubeMetadata } from '../../../lib/k8s/KubeMetadata';
+import { KubeObject } from '../../../lib/k8s/KubeObject';
+import { filterGraph, GraphFilter } from './graphFiltering';
+import { GraphEdge, GraphNode } from './graphModel';
+
+describe('filterGraph', () => {
+  const nodes: GraphNode[] = [
+    {
+      id: '1',
+      type: 'kubeObject',
+      data: {
+        resource: { metadata: { namespace: 'ns1', name: 'node1' } as KubeMetadata } as KubeObject,
+      },
+    },
+    {
+      id: '2',
+      type: 'kubeObject',
+      data: {
+        resource: {
+          kind: 'Pod',
+          metadata: { namespace: 'ns2' } as KubeMetadata,
+          status: { phase: 'Failed' },
+        } as any,
+      },
+    },
+    {
+      id: '3',
+      type: 'kubeObject',
+      data: { resource: { metadata: { namespace: 'ns3' } as KubeMetadata } as KubeObject },
+    },
+    {
+      id: '4',
+      type: 'kubeObject',
+      data: { resource: { metadata: { namespace: 'ns3' } as KubeMetadata } as KubeObject },
+    },
+  ];
+
+  const edges: GraphEdge[] = [
+    { id: 'e1', source: '1', target: '2', type: 'kubeRelation' },
+    { id: 'e2', source: '3', target: '4', type: 'kubeRelation' },
+  ];
+
+  it('filters nodes by name', () => {
+    const filters: GraphFilter[] = [{ type: 'name', query: 'node1' }];
+    const { nodes: filteredNodes } = filterGraph(nodes, edges, filters);
+    expect(filteredNodes.map(it => it.id)).toEqual(['1', '2']);
+  });
+
+  it('filters nodes by namespace', () => {
+    const filters: GraphFilter[] = [{ type: 'namespace', namespaces: new Set(['ns3']) }];
+    const { nodes: filteredNodes } = filterGraph(nodes, edges, filters);
+    expect(filteredNodes.map(it => it.id)).toEqual(['3', '4']);
+  });
+
+  it('filters nodes by related nodes', () => {
+    const filters: GraphFilter[] = [{ type: 'related', id: '1' }];
+    const { nodes: filteredNodes } = filterGraph(nodes, edges, filters);
+    expect(filteredNodes.map(it => it.id)).toEqual(['1', '2']);
+  });
+
+  it('filters nodes by custom filter function', () => {
+    const filters: GraphFilter[] = [
+      { type: 'custom', label: 'custom', filterFn: node => node.id === '1' },
+    ];
+    const { nodes: filteredNodes } = filterGraph(nodes, edges, filters);
+    expect(filteredNodes.map(it => it.id)).toEqual(['1', '2']);
```

Comment:
Can you please explain this test?

Maybe it's a bug. I would assume a filter for `id === '1'` would return only the id of '1'.

Reply:
filterGraph also includes related nodes, so this is expected. added a comment to both filterGraph and inside tests

I now realize the the "related" filter should have been named "id" filter instead but at this point this change can be implemented in later PR